Core: Speed up CollectionState sweeping

Mysteryem opened the pull request:
## What is this fixing or adding?

Increases the performance of sweeping state in `CollectionState.sweep_for_events()` and `MultiWorld.can_beat_game`, without changing the generated output (at least for supported games with their template yamls).

`CollectionState.sweep_for_events()` takes up a significant amount of generation time, mostly from filling progression items, and the majority when not generating a playthrough, so improvements to its performance can have a noticeable effect on generation duration.

Similarly, `MultiWorld.can_beat_game()` takes up a significant amount of playthrough calculation time from culling spheres.

This patch reduces the duration of both and refactors them to use the same sweeping code.

This patch has been split from #3688 to focus only on the sweep itself, and has been simplified to not require any additions to the existing world API.

### Make better use of region accessibility caching

Region accessibility is cached per-player and the cache for that player becomes stale whenever an item is collected or removed for that player. The original sweeping implementations make use of this by only collecting items once all reachable locations have been found for all players.

This patch changes to the sweep to find reachable locations per-player, collect the items from those locations and then move onto the next player. This keeps region accessibility cached as before, but has the advantage that if a player's reachable locations contain an item that gives a later player access to more locations, it will take fewer iterations for all reachable locations to be collected from.

### Skip checking location accessibility for locations belonging to a player which did not receive any advancement in the previous iteration

If no advancement items belonging to a player were collected in the previous sweep iteration (after the point at which their reachable locations were found), then it is assumed that player should have zero newly reachable locations, so checking for their reachable locations can be skipped.

While almost all worlds only logically depend on only their own items/locations etc., it is possible for a world to logically depend on a part of another world. To account for this, once the sweep runs out of accessible locations, the locations of all players are checked to confirm that the sweep is indeed finished.

My previous idea to account for worlds logically depending on other worlds was to require all worlds to specify which worlds they logically depended on. Performing extra sweeps to confirm will be slower than this previous idea, but it is simpler and does not require any changes to the existing world API.

## How was this tested?

Fixed seeds were generated, with and without this patch, with different sets of template yamls in the Players directory, using games known to have deterministic generation. The output from each set of template yamls was seen to match. Playthrough calculation is non-deterministic, so was not considered here.

Seeds using item links have also been tested.

To test the case of worlds depending on other worlds, I modified the A Hat in Time world to find all other AHiT worlds in the multiworld and then added additional rules to a few locations to require specific progression items belonging to one of those other AHiT worlds. I added some extra print statements to make sure that it would sometimes check if the sweep was finished (by checking every player), but then find newly reachable locations and then continue sweeping.

I also tried adding additional rules to a few entrances to require specific progression items belonging to a different world, but, unsurprisingly, this does not work well because entrance accessibility is cached per-player and only staled when the world the entrance belongs to collects an item, so the required item belonging to a different world could be collected, but without the entrance accessibility ever being staled to result in checking if the entrance is now accessible. This would often result in an error at the end of playthrough calculation because the end of playthrough calculation builds up locations by sphere, but `MultiWorld.can_beat_game()`, used to cull the spheres, no longer sweeps by spheres, so the two methods of sweeping could stale the cached entrance accessibility at different times, leading to the per-sphere sweep at the end of playthrough calculation to sometimes not be able to reach every required location.

---

Values in the table are rounded to 1 decimal place.
Note that generation times with every game included do have a small amount of variance in generation time because maybe 1 or 2 games have non-deterministic generation issues.
These times were recorded based on the main branch at f5218faea739 (before Kingdom Hearts was merged). As the logic in the main branch changes and is merged into this PR, the same seed may take longer or shorter to generate, so these times may not match the current state of main and this PR.

Generated with `python -O .\Generate.py --skip_output --seed 1` with a meta.yaml to disable progression balancing.
| Yamls | Before Patch/s | After Patch/s | Decrease in duration |
| -- | -- | -- | -- |
| 1 of each template YAML (67), averaged over 5 individual generations | 122.5 | 74.9 | 38.9% |
| 2 of each template YAML (134), averaged over 5 individual generations* | 371.6 | 192.0 | 48.3% |
*generated with `--seed 2` because `--seed 1` produced two sets of very different results, indicating a non-deterministic generation issue that caused two sets of very different seeds to be generated

Generated with `python -O .\Generate.py --seed 1`  with a meta.yaml to disable progression balancing, and `spoiler: 2` in `host.yaml`.
| Yamls | Before Patch/s | Only 'Make better use of region accessibility caching' | After Patch/s | Decrease in duration before -> after |
| -- | -- | -- | -- | -- |
| 1 of each non-rom template YAML (53), averaged over 5 individual generations | 504.5 | 436.6 | 376.4 | 25.4% |
| Above, but playthrough calculation only | 446.9 | 390.4 | 335.4 | 25.0% |
| Above, but without playthrough calculation | 57.6 | 46.2 | 41.0 | 28.7% |

Mysteryem commented:
The 3.8 ubuntu test failure appears to be unrelated to the changes in this PR, the zillion world's library appears to have been out-of-date which resulted in the world failing to load with ```OSError: pytest: reading from stdin while output is captured!  Consider using `-s`.```

NewSoupVi commented on BaseClasses.py:
```diff
@@ -754,20 +737,116 @@ def can_reach_entrance(self, spot: str, player: int) -> bool:
+                # If the player does not receive any items for the rest of the sweep iteration, they will likely not
+                # have any reachable locations at the start of the next sweep iteration, so their locations can be
+                # skipped.
+                next_players_to_check.discard(player)
```
I did some testing here on a hunch that `.discard` and `.add` on a set are slower than `dict[player] = True` and `dict[player] = False`.

This turned out to be true, but the useful thing you gain by using a set that you *remove* from is that you can **iterate it directly** and also check for its length to know when you're done.

The added overhead of checking the True or False from the dict for each player makes it slower.

So what you did is probably best

NewSoupVi commented:
I did not actually run the code, but it looks correct. Very awesome work

NewSoupVi reviewed (approved):
I think this is code that is sensitive enough to warrant approval from multiple core maintainers and should maybe wait to be merged until the start of a release cycle, but from my side, this is really good.

Mysteryem commented:
I've updated the description for the tests I did with A Hat in Time modified to add extra rules that depended on other worlds in the multiworld.

I don't think entrances with rules depending on items belonging to another world can be supported with the current caching for entrance accessibility being per-player, but this patch does increase how often such entrances will cause an error at the end of playthrough calculation.

> To test the case of worlds depending on other worlds, I modified the A Hat in Time world to find all other AHiT worlds in the multiworld and then added additional rules to a few locations to require specific progression items belonging to one of those other AHiT worlds. I added some extra print statements to make sure that it would sometimes check if the sweep was finished (by checking every player), but then find newly reachable locations and then continue sweeping.
>
> I also tried adding additional rules to a few entrances to require specific progression items belonging to a different world, but, unsurprisingly, this does not work well because entrance accessibility is cached per-player and only staled when the world the entrance belongs to collects an item, so the required item belonging to a different world could be collected, but without the entrance accessibility ever being staled to result in checking if the entrance is now accessible. This would often result in an error at the end of playthrough calculation because the end of playthrough calculation builds up locations by sphere, but `MultiWorld.can_beat_game()`, used to cull the spheres, no longer sweeps by spheres, so the two methods of sweeping could stale the cached entrance accessibility at different times, leading to the per-sphere sweep at the end of playthrough calculation to sometimes not be able to reach every required location.

beauxq commented on BaseClasses.py:
```diff
@@ -754,20 +737,116 @@ def can_reach_entrance(self, spot: str, player: int) -> bool:
+            # Create the generator, but tell it not to yield anything, so it will run to completion in zero iterations
+            # once started, then start and exhaust the generator by attempting to iterate it.
+            for _ in self._sweep_for_events_impl(events_per_player, False):
+                assert False, "Generator yielded when it should have run to completion without yielding"
```
```suggestion
                assert False, "Generator yielded when it should have run to completion without yielding"
            return None
```
PEP8 says this:
> Be consistent in return statements. Either all return statements in a function should return an expression, or none of them should. If any return statement returns an expression, any return statements where no value is returned should explicitly state this as return None, and an explicit return statement should be present at the end of the function (if reachable)

beauxq commented on BaseClasses.py:
```diff
@@ -754,20 +737,116 @@ def can_reach_entrance(self, spot: str, player: int) -> bool:
 
-    def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> None:
+    def _sweep_for_events_impl(self, events_per_player: List[Tuple[int, List[Location]]], yield_each_sweep: bool,
+                               ) -> Iterable[None]:
```
```suggestion
                               ) -> Iterator[None]:
```
The `Iterator` return type annotation is a common signal to people that this is an object that can only be iterated once - rather than some kind of collection, like a list or tuple, that can be iterated multiple times.

Mysteryem replied:
Looking at the docs, `Iterable` requires that `__iter__()` is implemented whereas `Iterator` requires that both `__iter__()` and `__next__()` are implemented, so `Iterator` is more accurate too.

Before I commit the change, is there any preference or reasons against specifying the return type even more accurately as `Generator[None, None, None]` instead?

beauxq replied:
I think I would find `Generator[None, None, None]` to be more noisy without much added value. But it's not bad if you really want it.

beauxq commented on BaseClasses.py:
```diff
@@ -754,20 +737,116 @@ def can_reach_entrance(self, spot: str, player: int) -> bool:
+                yield
+
+    def sweep_for_events(self, locations: Optional[Iterable[Location]] = None, yield_each_sweep: bool = False,
+                         checked_locations: Optional[Set[Location]] = None) -> Optional[Iterable[None]]:
```
```suggestion
                         checked_locations: Optional[Set[Location]] = None) -> Optional[Iterator[None]]:
```
same as above return type

NewSoupVi commented:
The entrance thing would kill the scheherazade apworld as well, but it sounds like it's already broken and just happens to work out sometimes? (I haven't thought that through yet)

I'm honestly wondering why you did end up deciding against the "dependent worlds" thing - It can be implemented in such a way that "only the worlds doing the funky thing" need to actually do anything, right? I think in main rn, that's only HK, and I think scheherazade would appreciate working with one extra line of code way more than it appreciates the efforts to make this generic

NewSoupVi commented:
Actually, I think HK doesn't even care, because it's checking for other players' items in can_beat_game, which shouldn't even be an issue

NewSoupVi commented:
Here's a different PR that uses the pattern of "attribute on the world class to change the behavior of sweep" https://github.com/ArchipelagoMW/Archipelago/pull/3682
And I quite like that one as well

Mysteryem commented:
> The entrance thing would kill the scheherazade apworld as well, but it sounds like it's already broken and just happens to work out sometimes? (I haven't thought that through yet)
> 
> I'm honestly wondering why you did end up deciding against the "dependent worlds" thing - It can be implemented in such a way that "only the worlds doing the funky thing" need to actually do anything, right? I think in main rn, that's only HK, and I think scheherazade would appreciate being able to exist flawlessly with one extra line of code way more than it appreciates the efforts to make this generic that don't quite help it as much as it needs
> 
> And for the record, depending on other worlds' stuff is absolutely a very special specific case.

Personally, the "dependent worlds" added a lot of bloat to the PR for handling a niche case. I stripped the original PR down to focus solely on the sweeping for this PR because it was the most significant part of the original PR. I considered not included any changes to `can_beat_game` in this PR, but I thought the conversion of `sweep_for_events` to be able to return a generator so that `can_beat_game` could use it would be better done in one go.

I think "dependent worlds" could still be introduced, but I think it would be better as its own PR because there are other areas that ideally would be changed too. The lookup of dependent players probably should be cached starting at a specific step of the generate process and I think it could also be used to stale region accessibility of dependent players, thus allowing for entrance logic to depend on other worlds. I haven't even started working towards a separate PR yet.

NewSoupVi commented:
> > The entrance thing would kill the scheherazade apworld as well, but it sounds like it's already broken and just happens to work out sometimes? (I haven't thought that through yet)
> > I'm honestly wondering why you did end up deciding against the "dependent worlds" thing - It can be implemented in such a way that "only the worlds doing the funky thing" need to actually do anything, right? I think in main rn, that's only HK, and I think scheherazade would appreciate being able to exist flawlessly with one extra line of code way more than it appreciates the efforts to make this generic that don't quite help it as much as it needs
> > And for the record, depending on other worlds' stuff is absolutely a very special specific case.
> 
> Personally, the "dependent worlds" added a lot of bloat to the PR for handling a niche case. I stripped the original PR down to focus solely on the sweeping for this PR because it was the most significant part of the original PR. I considered not included any changes to `can_beat_game` in this PR, but I thought the conversion of `sweep_for_events` to be able to return a generator so that `can_beat_game` could use it would be better done in one go.
> 
> I think "dependent worlds" could still be introduced, but I think it would be better as its own PR because there are other areas that ideally would be changed too. The lookup of dependent players probably should be cached starting at a specific step of the generate process and I think it could also be used to stale region accessibility of dependent players, thus allowing for entrance logic to depend on other worlds. I haven't even started working towards a separate PR yet.

Is it not as simple as having each world define `depends_on_slots: Optional[Set[int]]`, then collect that somewhere & flip it to `affects_slots: Dict[int, Set[int]]` (a lookup of players to worlds they affect, with the own slot being in the set by default)

and then you just replace `next_players_to_check.add(item.player)` with `next_players_to_check |= affects_worlds[item.player]`?
It probably introduces a small performance hit but I don't think it should be huge

NewSoupVi commented:
I mean, I say "just", it's obviously extra code. But yeah I think even as an extra PR, I don't imagine this will be enormous, barring any unforeseen problems I'm forgetting it's probably under 10 additional lines overall

Mysteryem commented:
> 
> Is it not as simple as having each world define `depends_on_slots: Optional[Set[int]]`, then collect that somewhere & flip it to `affects_slots: Dict[int, Set[int]]` (a lookup each slot to the slots they affect, with the own slot being in the set by default)
> 
> and then you just replace `next_players_to_check.add(item.player)` with `next_players_to_check |= affects_slots[item.player]`? It probably introduces a small performance hit but I don't think it should be huge
> 
> Edit: Oh, and the same change whereever regions get made stale ofc.

I'm pretty sure the 'flipping' process also needs to be recursive so that if world A's location logic uses can_reach on world B's location (A depends on B) and world B's location logic checks for items belonging to world C (B depends on C) that you end up with world A depending on both B and C. If the only part of worlds that other worlds could depend on were items then I don't think this would be necessary.

A problem I ran into when implementing caching for the lookup originally was that I think worlds would probably want to adjust which worlds they depend on as late as when they are allowed to set rules, but some worlds call `CollectionState.sweep_for_events()` earlier than that, e.g. Lingo appears to be the earliest with a call in `create_regions()`.

NewSoupVi commented:
> A problem I ran into when implementing caching for the lookup originally was that I think worlds would probably want to adjust which worlds they depend on as late as when they are allowed to set rules, but some worlds call `CollectionState.sweep_for_events()` earlier than that, e.g. Lingo appears to be the earliest with a call in `create_regions()`.

That's true, although those are singleplayer sweeps, right? There's probably some sort of assert you can do there, like

"this world defines dependent_slots, but swept for advancements with off-world locations before [cutoff point for setting dependent_slots] was reached"

NewSoupVi commented:
This PR is affected by the renaming of `sweep_for_events` to `sweep_for_advancements` and has conflicts now :)

Zannick commented on BaseClasses.py:
```diff
@@ -759,20 +742,117 @@ def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> No
+                    advancements_per_player.append((player, filtered_locations))
+        else:
+            # Filter and separate the locations into a list for each player.
+            advancements_per_player_dict: Dict[int, List[Location]] = defaultdict(list)
+            for location in locations:
+                if location.advancement and location not in checked_locations:
+                    advancements_per_player_dict[location.player].append(location)
+            # Convert to a list of tuples.
+            advancements_per_player = list(advancements_per_player_dict.items())
+            del advancements_per_player_dict
```
How often do you expect this to be only a small subset of all players? I would just directly construct one tuple for each player anyway.

```suggestion
            advancements_per_player: List[Tuple[int, List[Location]]] = [
                (player_id, []) for player_id in self.multiworld.player_ids
            ]
            for location in locations:
                if location.advancement and location not in checked_locations:
                    advancements_per_player[location.player - 1][1].append(location)
```

Mysteryem replied:
I expect it to be most of the time because the vast majority of sweeping does not use this conditional branch to begin with. Progression fill (`Fill.distribute_items_restrictive()` calling `Fill.fill_restrictive()`) and playthrough calculation (`Spoiler.create_playthrough()` calling `MultiWorld.can_beat_game()`), both pass in `locations=None` and thus do not run this conditional branch.

A world's `pre_fill()` often sweeps with only the locations belonging to the current player (`Fill.fill_restrictive(single_player_placement=True)`).
`Fill.distribute_early_items` also does a sweep limited to event (`location.address is None`) locations and only some games use events.
It looks like progression balancing also does some sweeping limited to locations belonging to a specific player, but I haven't looked into progression balancing in much detail yet.

Zannick reviewed (approved):
This looks very similar to work I did on OoTR many years ago, so the core dev chat suggested I take a look, and I think it looks great! :) Just a couple minor suggestions and questions.

I merged this and main into a test branch and ran `python Generate.py --skip_output --seed=42` with [line_profiler](https://github.com/pyutils/line_profiler) enabled on `sweep_for_advancements` in order to test this code and compare my suggestion below, on 11 templates (MM2 through Pokemon R&B). It exited with success (but obviously no output).

Mysteryem commented on BaseClasses.py:
```diff
@@ -759,20 +742,117 @@ def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> No
+            # Create the generator, but tell it not to yield anything, so it will run to completion in zero iterations
+            # once started, then start and exhaust the generator by attempting to iterate it.
+            for _ in self._sweep_for_advancements_impl(advancements_per_player, False):
+                assert False, "Generator yielded when it should have run to completion without yielding"
```
I would not have considered the performance difference to be ambiguous. Negligible? Yes. Ambiguous? No.

The `yield` will keep moving control back and forth between the generator's body and the code that is iterating the generator, whereas the conditional yield will not do this once, at the cost of checking a bool condition each loop.

Zannick replied:
Ok, wrong word choice. And I tested it out with `timeit` and found it to be much better to check whether to yield, as you have done. TIL

beauxq commented on BaseClasses.py:
```diff
@@ -759,20 +742,117 @@ def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> No
+            advancements_per_player = next_advancements_per_player
+
+            if yield_each_sweep:
+                yield
+
+    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, yield_each_sweep: bool = False,
```
```suggestion
                yield

    @typing.overload
    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, *,
                               yield_each_sweep: typing.Literal[True],
                               checked_locations: Optional[Set[Location]] = None) -> Iterator[None]: ...

    @typing.overload
    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None,
                               yield_each_sweep: typing.Literal[False] = False,
                               checked_locations: Optional[Set[Location]] = None) -> None: ...

    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, yield_each_sweep: bool = False,
```
This makes it so type-checkers and IDEs can see which calls return `None` and which calls return `Iterator`
so that it doesn't complain about returning an `Iterator` from `sweep_for_events`
or about iterating through `None` in `can_beat_game`

Mysteryem replied:
Thanks, I couldn't remember how to get the `yield_each_sweep: typing.Literal[True],` overload to work properly due to the default value of `False`.

NewSoupVi commented:
Just want to "bump" that while I approved this, I think another core approval would be good
